src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,181 @@
+            _disposalTokenSource.Token);
+    }
+
+    public void Dispose()
+    {
+        _disposalTokenSource.Cancel();
+        _disposalTokenSource.Dispose();
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> workItems, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            UpdateProject(workItem.ProjectKey, workItem.ProjectInfo, cancellationToken);
+        }
+
+        return default;
+    }
+
+    public void ProjectInfoUpdated(ProjectKey projectKey, RazorProjectInfo? projectInfo, CancellationToken cancellationToken)
+    {
+        var knownProject = _projectManager.TryGetLoadedProject(projectKey, out _);
+
+        if (projectInfo is null)
+        {
+            if (knownProject)
+            {
+                // projectInfo null is "remove"
+                EnqueueUpdateProject(projectKey, projectInfo: null);
+            }
+            else
+            {
+                _logger.LogWarning($"Found no existing project key '{projectKey.Id}' but projectInfo is null. Assuming no-op deletion.");
+            }
+
+            return;
+        }
+
+        if (!knownProject)
+        {
+            // For the new code path using endpoint, SerializedFilePath is IntermediateOutputPath.
+            // We will rename it when we remove the old code path that actually uses serialized bin file.
+            var intermediateOutputPath = FilePathNormalizer.Normalize(projectInfo.SerializedFilePath);
+            _logger.LogInformation($"Found no existing project key for project key '{projectKey.Id}'. Assuming new project.");
+
+            AddProjectAsync(
+                intermediateOutputPath,
+                projectInfo,
+                cancellationToken).Forget();
+            return;
+        }
+
+        _logger.LogInformation($"Project info changed for project '{projectKey.Id}'");
+
+        EnqueueUpdateProject(projectKey, projectInfo);
+    }
+
+    private async Task AddProjectAsync(
+        string intermediateOutputPath,
+        RazorProjectInfo projectInfo,
+        CancellationToken cancellationToken)
+    {
+        var projectFilePath = FilePathNormalizer.Normalize(projectInfo.FilePath);
+        var rootNamespace = projectInfo.RootNamespace;
+
+        var projectKey = await _projectService.AddProjectAsync(
+            projectFilePath,
+            intermediateOutputPath,
+            projectInfo.Configuration,
+            rootNamespace,
+            projectInfo.DisplayName,
+            cancellationToken).ConfigureAwait(false);
+
+        _logger.LogInformation($"Project configuration file added for project '{projectFilePath}': '{intermediateOutputPath}'");
+        EnqueueUpdateProject(projectKey, projectInfo);
```

Comment:
If the `ProjectInfoUpdated` method is changed per my above comment, this line could be deleted, and this method could just return the task from `AddProjectAsync` and avoid the await